contracts/services/QueryService.sol

```diff
@@ -363,4 +399,25 @@ contract QueryService is IVersioned {
             / (10 ** IToken(address(token)).decimals());
     }
 
+    function countAssets(
+        address issuer,
+        address investor,
+        address[] memory assetFactories
+    ) public view returns (uint256) {
+        if (assetFactories.length == 0) { return 0; }
+
+        // COUNT TOTAL RESPONSE SIZE
+        uint256 responseItemsCount = 0;
+        for (uint256 i = 0; i < assetFactories.length; i++) {
+            address[] memory instances = IAssetFactoryCommon(assetFactories[i]).getInstancesForIssuer(issuer);
+            for (uint256 j = 0; j < instances.length; j++) {
+                if (IToken(instances[j]).balanceOf(investor) > 0) { 
+                    responseItemsCount++;
+                }
+            }
+        }
```

Comment:
You're iterating through factories and their assets:
* once (here) - to get array size, and
* again (in the main method) - to make the response.

Isn't it double work that can be done in one pass? Consider that you're double-checking investors' balance in every single pass.
If you need to know the array size for structs ahead of time, can't you just create a map (`mapping(uint => Structs.AssetBalance)` where the keys would be array indices (1,2,3...), and then, in the end, you will end up with the map and the total number for items in the response. Then you could just create an array, and fill it with the data in the loop over map items. Does it make sense? Or is your approach really more efficient to double iterate to save memory allocation frequency? We can discuss this in person if you like.